### ISSUE TITLE
wrong type inference for square terms

### ISSUE THREAD

**Issue by githubtomtom** [JuliaStats/StatsModels.jl#244]
[please read this](https://discourse.julialang.org/t/bug-in-statsmodels-modelmatrix/69428/2).

Currently, formulas like `@formula(y ~ 1 + a * (a + b) + b * b)` would fail. We **need to explicitly state the square terms** like `@formula(y ~ 1 + a + (a ^ 2) + (a & b) + b + (b ^ 2))`.

It's because now term like `a & a` would has type `InteractionTerm{ContinuousTerm{Float64}}` rather than `InteractionTerm{Tuple{ContinuousTerm{Float64}, ContinuousTerm{Float64}}}`.

Note that currently `InteractionTerm{ContinuousTerm{Float64}}` can **not** be properly handled.

**Comment by kleinschmidt** [comment]
Yup, if you want `a .^ 2` then `a ^ 2` is the way to do it.  If you want both `a` and `a^2`, that starts to sound more like a polynomial basis set...the idea with StatsModels is that users/developers can create their own packages that, when loaded, provide special syntax to do things like that, and a few people have kicked around the idea of making a "basis functions" package that provides that interface. 

As a toy example, you might also be interested in looking at the example from the docs about [how to create a "poly term"](https://juliastats.org/StatsModels.jl/stable/internals/#extending) that generates [a, a^2, a^3, ..., a^n] from `@formula(0 ~ poly(a, n))`, and lets you do things like

```julia
julia> fit(LinearModel, @formula(y ~ 1 + poly(a,2) + poly(b,2)), sim_dat)
StatsModels.TableRegressionModel{LinearModel{GLM.LmResp{Array{Float64,1}},GLM.DensePredChol{Float64,LinearAlgebra.Cholesky{Float64,Array{Float64,2}}}},Array{Float64,2}}

y ~ 1 + poly(a, 2) + poly(b, 2)

Coefficients:
──────────────────────────────────────────────────────────────────────────
                 Coef.  Std. Error      t  Pr(>|t|)   Lower 95%  Upper 95%
──────────────────────────────────────────────────────────────────────────
(Intercept)   0.89288    0.181485    4.92    <1e-5    0.532586    1.25317
a^1           2.73324    0.349194    7.83    <1e-11   2.04001     3.42648
a^2          -1.0114     1.34262    -0.75    0.4531  -3.67684     1.65404
b^1           0.214424   0.136868    1.57    0.1205  -0.0572944   0.486142
b^2           3.15133    0.0811794  38.82    <1e-59   2.99016     3.31249
──────────────────────────────────────────────────────────────────────────
```
